src/librustc/hir/lowering.rs

```diff
@@ -2389,7 +2391,10 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
     ) -> hir::GenericBound {
         // Compute the `T` in `Future<Output = T>` from the return type.
         let output_ty = match output {
-            FunctionRetTy::Ty(ty) => self.lower_ty(ty, ImplTraitContext::OpaqueTy(Some(fn_def_id))),
+            FunctionRetTy::Ty(ty) => self.lower_ty(
+                ty,
+                ImplTraitContext::OpaqueTy(Some(fn_def_id), hir::OpaqueTyOrigin::FnReturn),
```

Comment:
Not `OpaqueTyOrigin::AsyncFn`?

Reply:
No, this is for user specified `impl Trait` in the return type of an `async fn`

Reply:
Oh; guess that might be clearer had I read more of the code, but maybe leave a comment?